File: dialects/f5/bigip/types/_snat_mode.py
"""Typed BIG-IP source-address-translation mode value.

The ``source-address-translation`` property on an ltm virtual (and
the related ``snatpool`` field) is a sum type with four forms:

    source-address-translation { type none }
    source-address-translation { type automap }
    source-address-translation { type snat pool /Common/snatpool_a }
    snatpool /Common/snatpool_a

This module models the sum type so consumers can ask
``is this automap?``, ``which snat pool does this use?``, ``is
SNAT off here?`` without re-parsing the body, and the rewrite
layer can swap modes safely (an automap → pool conversion should
rewrite the body, not just splice a new string in).

The matching :class:`dialects.f5.bigip.registry.value_specs.SnatModeSpec`
wires the value into the registry.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

SnatModeKind = Literal["none", "automap", "snat", "address-list"]


@dataclass(frozen=True, slots=True)
class SnatMode:
    """A parsed source-address-translation mode.

    ``kind`` picks the variant:

    - ``"none"`` — SNAT is disabled (``type none``).
    - ``"automap"`` — BIG-IP auto-selects a self-IP per pool member
      (``type automap``).
    - ``"snat"`` — A named SNAT pool is used (``type snat`` + ``pool
      <full-path>``); ``pool_path`` holds the reference.
    - ``"address-list"`` — Reserved for the address-list variant
      some classifier objects use; ``addresses`` carries the list.

    ``raw`` preserves the original brace-body text so a no-op round
    trip through the rewrite layer doesn't change spacing.
    """

    kind: SnatModeKind = "none"
    pool_path: str = ""
    addresses: tuple[str, ...] = ()
    raw: str = ""

    @classmethod
    def try_parse(cls, text: str) -> "SnatMode | None":
        """Parse a ``source-address-translation`` body.

        Accepts both shapes the parser layer might hand in:

        - The full property body ``{ type snat pool /Common/p }`` (with
          braces).
        - The inner content ``type snat pool /Common/p`` (without
          braces, which is how :func:`_parse_properties_with_spans`
          reports it after stripping the outer ``{`` / ``}``).
        """
        if text is None:
            return None
        body = text.strip()
        if not body:
            return None
        if body.startswith("{"):
            body = body[1:]
        if body.endswith("}"):
            body = body[:-1]
        tokens = body.split()
        if not tokens:
            return None
        kind: SnatModeKind = "none"
        pool_path = ""
        idx = 0
        while idx < len(tokens):
            tok = tokens[idx]
            if tok == "type" and idx + 1 < len(tokens):
                t = tokens[idx + 1]
                if t == "none":
                    kind = "none"
                elif t == "automap":
                    kind = "automap"
                elif t == "snat":
                    kind = "snat"
                else:
                    return None
                idx += 2
                continue
            if tok == "pool" and idx + 1 < len(tokens):
                pool_path = tokens[idx + 1]
                idx += 2
                continue
            idx += 1
        if kind == "snat" and not pool_path:
            # ``type snat`` without a pool reference is malformed.
            return None
        return cls(kind=kind, pool_path=pool_path, raw=text.strip())

    @property
    def is_none(self) -> bool:
        return self.kind == "none"

    @property
    def is_automap(self) -> bool:
        return self.kind == "automap"

    @property
    def is_snat(self) -> bool:
        return self.kind == "snat"

    def references(self) -> tuple[str, ...]:
        """Return the full-paths this mode references (the SNAT pool,
        when the mode is ``snat``).  Empty for none / automap."""
        if self.kind == "snat" and self.pool_path:
            return (self.pool_path,)
        return ()

    def __str__(self) -> str:
        if self.raw:
            return self.raw
        if self.kind == "none":
            return "{ type none }"
        if self.kind == "automap":
            return "{ type automap }"
        if self.kind == "snat":
            return f"{{ type snat pool {self.pool_path} }}"
        return ""


__all__ = ["SnatMode", "SnatModeKind"]

File: rust/tcl-bigip/src/value/snat_mode.rs
//! Typed BIG-IP source-address-translation mode value.

use std::fmt;

/// The variant of a [`SnatMode`].
///
/// [`SnatModeKind::as_str`] returns the canonical string literal.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum SnatModeKind {
    /// SNAT disabled (`type none`).
    None,
    /// Auto-selected self-IP (`type automap`).
    Automap,
    /// Named SNAT pool (`type snat` + `pool <full-path>`).
    Snat,
    /// Reserved address-list variant.
    AddressList,
}

impl SnatModeKind {
    /// The canonical string literal for this kind.
    #[must_use]
    pub fn as_str(self) -> &'static str {
        match self {
            SnatModeKind::None => "none",
            SnatModeKind::Automap => "automap",
            SnatModeKind::Snat => "snat",
            SnatModeKind::AddressList => "address-list",
        }
    }
}

/// A parsed source-address-translation mode.
///
/// `raw` preserves the original brace-body text so a no-op round trip
/// doesn't change spacing.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct SnatMode {
    /// The mode variant.
    pub kind: SnatModeKind,
    /// SNAT pool reference (for the `snat` kind).
    pub pool_path: String,
    /// Address list (for the `address-list` kind).
    pub addresses: Vec<String>,
    /// Original body text (stripped).
    pub raw: String,
}

impl Default for SnatMode {
    fn default() -> Self {
        Self {
            kind: SnatModeKind::None,
            pool_path: String::new(),
            addresses: Vec::new(),
            raw: String::new(),
        }
    }
}

impl SnatMode {
    /// Parse a `source-address-translation` body, with or without the
    /// surrounding braces. Returns `None` for empty input or a malformed
    /// `type snat` clause without a pool.
    #[must_use]
    pub fn try_parse(text: &str) -> Option<Self> {
        let trimmed = text.trim();
        if trimmed.is_empty() {
            return None;
        }
        let mut body = trimmed;
        if let Some(rest) = body.strip_prefix('{') {
            body = rest;
        }
        if let Some(rest) = body.strip_suffix('}') {
            body = rest;
        }
        let tokens: Vec<&str> = body.split_whitespace().collect();
        if tokens.is_empty() {
            return None;
        }
        let mut kind = SnatModeKind::None;
        let mut pool_path = String::new();
        let mut idx = 0;
        while idx < tokens.len() {
            let tok = tokens[idx];
            if tok == "type" && idx + 1 < tokens.len() {
                match tokens[idx + 1] {
                    "none" => kind = SnatModeKind::None,
                    "automap" => kind = SnatModeKind::Automap,
                    "snat" => kind = SnatModeKind::Snat,
                    _ => return None,
                }
                idx += 2;
                continue;
            }
            if tok == "pool" && idx + 1 < tokens.len() {
                tokens[idx + 1].clone_into(&mut pool_path);
                idx += 2;
                continue;
            }
            idx += 1;
        }
        if kind == SnatModeKind::Snat && pool_path.is_empty() {
            return None;
        }
        Some(Self {
            kind,
            pool_path,
            addresses: Vec::new(),
            raw: trimmed.to_owned(),
        })
    }

    /// `true` for the `none` kind.
    #[must_use]
    pub fn is_none(&self) -> bool {
        self.kind == SnatModeKind::None
    }

    /// `true` for the `automap` kind.
    #[must_use]
    pub fn is_automap(&self) -> bool {
        self.kind == SnatModeKind::Automap
    }

    /// `true` for the `snat` kind.
    #[must_use]
    pub fn is_snat(&self) -> bool {
        self.kind == SnatModeKind::Snat
    }

    /// Return the full-paths this mode references (the SNAT pool, for
    /// `snat`). Empty otherwise.
    #[must_use]
    pub fn references(&self) -> Vec<String> {
        if self.kind == SnatModeKind::Snat && !self.pool_path.is_empty() {
            return vec![self.pool_path.clone()];
        }
        Vec::new()
    }
}

impl fmt::Display for SnatMode {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if !self.raw.is_empty() {
            return f.write_str(&self.raw);
        }
        match self.kind {
            SnatModeKind::None => f.write_str("{ type none }"),
            SnatModeKind::Automap => f.write_str("{ type automap }"),
            SnatModeKind::Snat => write!(f, "{{ type snat pool {} }}", self.pool_path),
            SnatModeKind::AddressList => Ok(()),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_forms() {
        let m = SnatMode::try_parse("type none").unwrap();
        assert!(m.is_none());
        let m = SnatMode::try_parse("{ type automap }").unwrap();
        assert!(m.is_automap());
        assert_eq!(m.to_string(), "{ type automap }");
        let m = SnatMode::try_parse("type snat pool /Common/snatpool_a").unwrap();
        assert!(m.is_snat());
        assert_eq!(m.pool_path, "/Common/snatpool_a");
        assert_eq!(m.references(), vec!["/Common/snatpool_a".to_owned()]);
    }

    #[test]
    fn structured_render() {
        let m = SnatMode {
            kind: SnatModeKind::Snat,
            pool_path: "/Common/p".to_owned(),
            ..Default::default()
        };
        assert_eq!(m.to_string(), "{ type snat pool /Common/p }");
        let m = SnatMode::default();
        assert_eq!(m.to_string(), "{ type none }");
    }

    #[test]
    fn rejects_bad() {
        assert!(SnatMode::try_parse("").is_none());
        assert!(SnatMode::try_parse("type snat").is_none());
        assert!(SnatMode::try_parse("type bogus").is_none());
    }
}
